File: backend/app/schemas/event_content.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.schemas.periods import PresetPeriod

EventContentPeriod = PresetPeriod | Literal["all"]
EventReleaseStatus = Literal["released", "draft", "review", "cancelled", "deferred", "rescheduled"]


class EventContentFilters(BaseModel):
    model_config = {"extra": "forbid"}
    period: EventContentPeriod = "24h"
    status: EventReleaseStatus | None = None
    compare: Literal["previous"] | None = None

    @model_validator(mode="after")
    def bounded_comparison(self) -> "EventContentFilters":
        if self.period == "all" and self.compare:
            raise ValueError("All has no previous period")
        return self


class AssignmentCoverage(BaseModel):
    events_with_assignment: int = Field(ge=0)
    events_without_assignment: int = Field(ge=0)
    coverage_percent: float = Field(ge=0, le=100)


class EventContentCoverage(BaseModel):
    categories: AssignmentCoverage
    genres: AssignmentCoverage
    event_types: AssignmentCoverage


class EventContentRankingItem(BaseModel):
    id: str
    name: str
    event_count: int = Field(ge=0)
    event_share_percent: float = Field(ge=0, le=100)
    rank: int = Field(ge=1)
    previous_rank: int | None = None
    rank_delta: int | None = None
    previous_event_count: int | None = None
    count_delta: int | None = None
    previous_share_percent: float | None = None
    share_delta_percentage_points: float | None = None


class EventContentRanking(BaseModel):
    distinct_assignment_count: int = Field(ge=0)
    items: list[EventContentRankingItem] = Field(max_length=10)


class EventContentComparison(BaseModel):
    from_at: datetime
    to_at: datetime
    event_count: int = Field(ge=0)
    coverage: EventContentCoverage


class EventContentStatistics(BaseModel):
    period: EventContentPeriod
    from_at: datetime | None
    to_at: datetime | None
    observed_at: datetime
    timezone: str
    status: EventReleaseStatus | None
    event_count: int = Field(ge=0)
    coverage: EventContentCoverage
    categories: EventContentRanking
    genres: EventContentRanking
    event_types: EventContentRanking
    comparison: EventContentComparison | None

File: frontend/tests/fixtures/event-content.ts
import type { EventContentStatistics, EventContentRanking } from '../../shared/contracts'
export function eventContentFixture(params = new URLSearchParams()): EventContentStatistics {
  const period = (params.get('period') ?? '24h') as EventContentStatistics['period']
  const status = params.get('status') as EventContentStatistics['status']
  const total = status === 'cancelled' ? 0 : status === 'released' ? 8 : 10
  const compare = params.get('compare') === 'previous' && period !== 'all'
  const end = '2026-09-17T12:00:00Z'
  const days = { today: 14 / 24, '24h': 1, '7d': 7, '30d': 30, '90d': 90, all: 1 }[period]
  const start = new Date(Date.parse(end) - days * 86400000).toISOString()
  function coverage(count: number, denominator: number) {
    return {
      events_with_assignment: count,
      events_without_assignment: denominator - count,
      coverage_percent: denominator ? (count / denominator) * 100 : 0,
    }
  }
  const ranking = (id: string, name: string): EventContentRanking => ({
    distinct_assignment_count: total ? 1 : 0,
    items: total
      ? [
          {
            id,
            name,
            event_count: 6,
            event_share_percent: (6 / total) * 100,
            rank: 1,
            previous_rank: compare ? 3 : null,
            rank_delta: compare ? 2 : null,
            previous_event_count: compare ? 4 : null,
            count_delta: compare ? 2 : null,
            previous_share_percent: compare ? 40 : null,
            share_delta_percentage_points: compare ? (6 / total) * 100 - 40 : null,
          },
        ]
      : [],
  })
  const categories = ranking('1', 'Musik')
  const eventTypes = ranking('1', 'Konzert')
  if (total) {
    for (const [data, name, count] of [
      [categories, 'Familie', 2],
      [eventTypes, 'Workshop', total - 6],
    ] as const) {
      data.distinct_assignment_count = 2
      data.items.push({
        ...data.items[0]!,
        id: '2',
        name,
        rank: 2,
        event_count: count,
        event_share_percent: (count / total) * 100,
        previous_rank: compare ? 4 : null,
        rank_delta: compare ? 2 : null,
        previous_event_count: compare ? 2 : null,
        count_delta: compare ? count - 2 : null,
        previous_share_percent: compare ? 20 : null,
        share_delta_percentage_points: compare ? (count / total) * 100 - 20 : null,
      })
    }
  }
  return {
    period,
    status,
    from_at: period === 'all' ? null : start,
    to_at: period === 'all' ? null : end,
    timezone: 'Europe/Berlin',
    observed_at: end,
    event_count: total,
    coverage: {
      categories: coverage(total ? 8 : 0, total),
      genres: coverage(total ? 6 : 0, total),
      event_types: coverage(total, total),
    },
    categories,
    genres: ranking('1:4', 'Konzert · Rock'),
    event_types: eventTypes,
    comparison: compare
      ? {
          from_at: new Date(Date.parse(start) - days * 86400000).toISOString(),
          to_at: start,
          event_count: 10,
          coverage: {
            categories: coverage(8, 10),
            genres: coverage(6, 10),
            event_types: coverage(9, 10),
          },
        }
      : null,
  }
}
